feat(tree): Implement TreeView v2 API (#20815)
microsoft/FluidFramework: .changeset/silly-clocks-tan.md

---
"@fluidframework/tree": major
---

Implement compatibility-based schema evolution API

> This change adjusts some top-level APIs for using SharedTree in a context that may necessitate changing a document's schema.

Reviewer: There are two grammatically valid ways to parse this sentence with different meanings. One implies that the adjustment of the API may necessitate changing a document's schema: this is the kind of thing we commonly see in changesets but not what this is saying. so I think it should be rephrased.

```suggestion
This change adjusts some top-level APIs for using SharedTree to better accommodate applications that need to change their schema.
```

Reviewer 2: By "there are two grammatically valid ways to parse this sentence with different meanings," are you trying to say that the sentence can be parsed in two ways, and each way has a different meaning? Or are you trying to say that the sentence, after being given different meanings, can be parsed in two ways?

> The motivation for these changes is to allow future relaxation of `SharedTree`'s restrictions around view schema and stored schema compatibility,
> which will enable more flexible policies around how applications can update their documents' schemas over time.

Reviewer: I think its better to lead with what changed than why in a changeset.
```suggestion
These changes enable forwards compatibility with future work to relax `SharedTree`'s restrictions around view schema and stored schema compatibility;
that future work will enable more flexible policies around how applications can update their documents' schemas over time.
```

Application authors are encouraged to develop a compatibility policy which they are comfortable with using the guidance in the
"Schema Evolvability" section of `@fluidframework/tree`'s readme.

To make the details of schema compatibilities that SharedTree supports more clear,
`TreeView.error` has been functionally replaced with the `compatibility` property.
Users desiring the previous strict behavior should use `view.compatibility.isEquivalent` at appropriate places in application logic.

# `ITree.schematize` deprecation

`ITree.schematize` has been deprecated in favor of `ITree.viewWith`.
Unlike `schematize`, `viewWith` does not implicitly initialize the document. As such, it doesn't take an `initialTree` property.
Instead, applications should initialize their trees in document creation codepaths using the added `TreeView.initialize` API.

As an example, something like the following code may have been used before for both the document create and document load codepaths:

```typescript

> const tree = SharedTree.create(runtime, "foo");

Reviewer: Use of the encapsulated API which is not public seems non-ideal (at least as the only example).

It might be nice to include an example of how we expect users of container schema to use the new API. (maybe both the static/in schema case and also the dynamic DDS creation case)

This change adjusts some top-level APIs for using SharedTree in a context that may necessitate changing a document's schema.
First of all, `SharedTree`'s restrictions around view schema and stored schema compatibility have been slightly relaxed:
A particular view schema can now be used to open documents whose stored schema supports a superset of that view schema.
Beware that applications which choose to do this may experience runtime incompatibilities in application logic: for instance,
the stored schema for a document may contain allowed types in fields which the view schema isn't expecting,
which could result in application errors.
The motivation for this change is to eventually allow more flexible application policies around what types of schema incompatibilities
are OK to allow different clients to collaborate using.

Application authors are encouraged to develop a compatibility policy which they are comfortable with using the guidance in the
"Schema Evolvability" section of `@fluidframework/tree`'s readme.

To make the details of schema compatibilities that SharedTree supports more clear,
`TreeView.error` has been functionally replaced with the `compatibility` property.

> Users desiring the previous strict behavior can use `view.compatibility.isExactMatch` at appropriate places of application logic.

Reviewer 3: Nit
```suggestion
Users desiring the previous strict behavior can use `view.compatibility.isExactMatch` at appropriate places in application logic.
```

In addition to this functional change to the types of collaboration that `SharedTree` permits, several APIs have been tweaked.

# `ITree.schematize` deprecation

`ITree.schematize` has been deprecated in favor of `ITree.viewWith`.
Unlike `schematize`, `viewWith` does not implicitly initialize the document. As such, it doesn't take an `initialTree` property.
Instead, applications should initialize their trees in document creation codepaths using the added `TreeView.initialize` API.

As an example, something like the following code may have been used before for both the document create and document load codepaths:

```typescript
const tree = SharedTree.create(runtime, "foo");

> const view = tree.schematize({ schema: Point, initialTree: () => new Point({ x: 0, y: 0 }) });

Reviewer: This would not work in the before, because it needs to construct a TreeConfiguration (its a class not a interface).

Having this before version be inaccurate in this way makes it hard to evaluate the intended change on applications. Currently apps tend to declare their configuration in the schema file, and the choice of the class based API was made to help them be type safe and improve runtime and compile time errors. I cannot tell if that is regressed by this change as the before and after APis for users like this are not clear.

Author: ```suggestion
const view = tree.schematize(new TreeConfiguration(Point, () => new Point({ x: 0, y: 0 })));
```

# `ITree.schematize` deprecation

`ITree.schematize` has been deprecated in favor of `ITree.viewWith`.
Unlike `schematize`, `viewWith` does not implicitly initialize the document.
As such, it doesn't take an `initialTree` property.
Instead, applications should initialize their trees in document creation codepaths using the added `TreeView.initialize` API.

## Old

As an example, something like the following code may have been used before for both the document create and document load codepaths:

```typescript
// -- first-party API --
const tree = SharedTree.create(runtime, "foo");
const view = tree.schematize(new TreeConfiguration(Point, () => new Point({ x: 0, y: 0 })));
```

When using the third-party API, creating a tree looks a bit different but the call to `schematize` is the same:

```typescript

> // -- third-party API for statically defined objects in container schema --

Reviewer: Given that we prefer first parties to use this API if practical, and third parties can technically do whatever they want (we don't actually have any special first party access), I don't think thats a great way to describe this API surface. I think any of `fluid-framework`, "declarative", `ContainerSchema`, "Service Client" or `IFluidContainer`.

Also I'd suggest putting this API first as its the public one we want people to use.

What you have called the "first party API" is called the `encapsulated API`, but its also exported under /legacy, you calling it the "legacy" or "/legacy" API would also make sense ( https://fluidframework.com/docs/build/releases-and-apitags/calls it legacy, though is incorrect about how its tagged).

https://github.com/microsoft/FluidFramework/wiki/Release-Tags uses the term "known partners" not first-party.
I think we should avoid introducing anything that claims to be special first party stuff. For developers outside MS contributing to this repo, first party likely won't mean what I think you are intending here.
